File: apps/macos/Sources/WispProjectBrowser/NativeShare.swift
import Foundation

public struct NativeShareRow: Codable, Equatable, Sendable {
    public var role: String
    public var text: String
}
public struct NativeShareDraftRow: Identifiable, Equatable {
    public let id: Int
    public var row: NativeShareRow
    public var selected: Bool
}
public enum NativeShare {
    public static func draft(_ rows: [NativeShareRow]) -> [NativeShareDraftRow] {
        rows.enumerated().compactMap { index, row in
            guard ["user", "assistant", "reasoning"].contains(row.role), !row.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
            return NativeShareDraftRow(id: index, row: row, selected: row.role != "reasoning")
        }
    }
    public static func keywords(_ raw: String) -> [String] {
        Array(Set(raw.components(separatedBy: CharacterSet(charactersIn: ",，\n")).map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }))
            .sorted { $0.count == $1.count ? $0 < $1 : $0.count > $1.count }
    }
    public static func redact(_ text: String, keywords: [String]) -> String {
        keywords.reduce(text) { result, word in result.replacingOccurrences(of: word, with: "xxx", options: .caseInsensitive) }
    }
    public static func width(_ raw: String) -> Int { min(2400, max(320, UInt(raw.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap(Int.init(exactly:)) ?? 840)) }
    public static func selected(_ draft: [NativeShareDraftRow], keywords: String) -> [NativeShareRow] {
        let words = self.keywords(keywords)
        return draft.filter(\.selected).map { NativeShareRow(role: $0.row.role, text: redact($0.row.text, keywords: words)) }
    }
}

File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeShare.cs
using System.Text.Json;
using System.Text.Json.Nodes;
namespace Wisp.ProjectBrowser.Contracts;
public sealed record NativeShareRow(string Role, string Text);
public sealed record NativeShareDraftRow(int Id, NativeShareRow Row, bool Selected);
public static class NativeShareDraft
{
    public static NativeShareDraftRow[] From(IEnumerable<NativeShareRow> rows) => rows.Select((row, index) => (row, index))
        .Where(pair => pair.row.Role is "user" or "assistant" or "reasoning" && !string.IsNullOrWhiteSpace(pair.row.Text))
        .Select(pair => new NativeShareDraftRow(pair.index, pair.row, pair.row.Role != "reasoning")).ToArray();
    public static string[] Keywords(string raw) => raw.Split(',', '，', '\n')
        .Select(word => word.Trim()).Where(word => word.Length > 0).Distinct().OrderByDescending(word => word.Length).ThenBy(word => word).ToArray();
    public static string Redact(string text, IEnumerable<string> keywords) =>
        keywords.Aggregate(text, (current, word) => current.Replace(word, "xxx", StringComparison.OrdinalIgnoreCase));
    public static int Width(string raw) => int.TryParse(raw.Trim(), out var value) ? Math.Min(2400, Math.Max(320, value)) : 840;
    public static NativeShareRow[] Selected(IEnumerable<NativeShareDraftRow> draft, string keywords)
    {
        var words = Keywords(keywords);
        return draft.Where(row => row.Selected).Select(row => row.Row with { Text = Redact(row.Row.Text, words) }).ToArray();
    }
}
/// <summary>Send only the user's selected, edited and redacted rows to HtmlAsync.
/// PNG rendering and file dialogs belong to the native frontend.</summary>
public interface INativeShareClient
{
    Task<NativeShareRow[]> ReadAsync(string projectId, string sessionId, CancellationToken cancellationToken = default);
    Task<string> HtmlAsync(string projectId, string sessionId, NativeShareRow[] rows, bool dark = false, CancellationToken cancellationToken = default);
}
public sealed class NativeShareClient(INativeSettingsClient transport) : INativeShareClient
{
    public async Task<NativeShareRow[]> ReadAsync(string projectId, string sessionId, CancellationToken cancellationToken = default) =>
        (await transport.InvokeAsync("native_conversation_share", new() { ["session_id"] = sessionId }, projectId, cancellationToken).ConfigureAwait(false))?.Deserialize<NativeShareRow[]>(ConversationSnapshot.JsonOptions)
            ?? throw new InvalidDataException("Missing share rows");
    public async Task<string> HtmlAsync(string projectId, string sessionId, NativeShareRow[] rows, bool dark = false, CancellationToken cancellationToken = default) =>
        (await transport.InvokeAsync("native_conversation_share_html", new() { ["session_id"] = sessionId, ["rows"] = JsonSerializer.SerializeToNode(rows, ConversationSnapshot.JsonOptions), ["dark"] = dark }, projectId, cancellationToken).ConfigureAwait(false))?.GetValue<string>()
            ?? throw new InvalidDataException("Missing share document");
}
